controllers/nginxingresscontroller_controller.go

```diff
@@ -0,0 +1,329 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *NginxIngressControllerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("nginxingresscontroller", req.NamespacedName)
+	log.Info("Reconciling NginxIngressController")
+
+	instance := &k8sv1alpha1.NginxIngressController{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil && errors.IsNotFound(err) {
+		// Request object not found, could have been deleted after reconcile request.
+		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+		// Return and don't requeue
+		log.Info("NginxIngressController resource not found. Ignoring since object must be deleted")
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to get NginxIngressController")
+		return ctrl.Result{}, err
+	}
+
+	// Check if the NginxIngressController instance is marked to be deleted, which is
+	// indicated by the deletion timestamp being set.
+	isNginxIngressControllerMarkedToBeDeleted := instance.GetDeletionTimestamp() != nil
+	if isNginxIngressControllerMarkedToBeDeleted {
+		if controllerutil.ContainsFinalizer(instance, finalizer) {
+			// Run finalization logic for nginxingresscontrollerFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			if err := r.finalizeNginxIngressController(log, instance); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove nginxingresscontrollerFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			controllerutil.RemoveFinalizer(instance, finalizer)
+			err := r.Update(ctx, instance)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Add finalizer for this CR
+	if !controllerutil.ContainsFinalizer(instance, finalizer) {
+		controllerutil.AddFinalizer(instance, finalizer)
+		err = r.Update(ctx, instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Namespace could have been deleted in the middle of the reconcile
+	ns := &v1.Namespace{}
+	err = r.Get(ctx, types.NamespacedName{Name: instance.Namespace, Namespace: v1.NamespaceAll}, ns)
+	if (err != nil && errors.IsNotFound(err)) || (ns.Status.Phase == "Terminating") {
+		log.Info(fmt.Sprintf("The namespace '%v' does not exist or is in Terminating status, canceling Reconciling", instance.Namespace))
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Error(err, "Failed to check if namespace exists")
+		return ctrl.Result{}, err
+	}
+
+	if err := r.createCommonResources(log, instance); err != nil {
```

Comment:
I think we should always check that everything that we need is in place, before we were checking only some of the resources. Ideally, we should refactor to use `controllerutil.CreateOrUpdate` where possible.

Reply:
once the operator creates those common resources, it doesn't monitor them - if those are changed or removed manually, the operator doesn't do anything. Operator assumed that those resources are not removed/changed.

If we want to prevent those possible changes/removals, I think we should have a separate reconcile loop that monitors those resources and restores them if those were changed. Note that the current reconcile loop will not be triggered if those resources were changed.  

Because it is not required for the sdk upgrade, I suggest doing in a separate pull request.

Reply:
how is this different than calling `checkPrerequisites` on every loop? Also calling it during setup fails, so not sure where else I could call it.

Reply:
> how is this different than calling checkPrerequisites on every loop? 

(a)
if the problem we're trying to solve is to prevent those resources from being modified, then calling `checkPrerequisites` will not work.  Let me explain why:

Let's say some of those common resources is modified: an RBAC ClusterRole is removed, for example.  If that is the case, the deployed Ingress Controllers will stop working correctly. However, after the removal of that ClusterRole, the `Reconcile` will not be called, so the Operator will not fix the Ingress Controllers. The `Reconcile` only will be called when any of the resources of any Ingress Controller are modified:
```
		For(&k8sv1alpha1.NginxIngressController{}).
		Owns(&appsv1.Deployment{}).
		Owns(&appsv1.DaemonSet{}).
		Owns(&v1.ServiceAccount{}).
		Owns(&v1.Service{}).
		Owns(&v1.ConfigMap{}).
		Owns(&v1.Secret{}).
```

Also, making the Operator ensure the common resources are always restored is the change of the existing logic. Is it required for SDK upgrade?  Shall  we minimize the number of changes and focus on upgrading the SDK only? The more changes we make, the more bugs we can potentially bring and it also makes it more difficult to review, because of the increased size/scope.
 

(b)
> Also calling it during setup fails, so not sure where else I could call it. 

It looks like you were trying to solve this specific problem.  Can we address that failure rather than changing the operator logic?

Reply:
I'm not trying to solve any problem. `checkPrerequisites` was already called in the loop and has a similar logic to `createCommonResources`, so I don't understand why you're saying that the first one can be in the loop and the second can't. 

You want me to move back `createCommonResources` out of the loop, right? Can you tell me where you think it should go? Calling it from `SetupWithManager` doesn't work because the resources are not ready yet.

Reply:
> I'm not trying to solve any problem. checkPrerequisites was already called in the loop and has a similar logic to createCommonResources, so I don't understand why you're saying that the first one can be in the loop and the second can't.

`checkPrerequisites` are prerequisites for a single Ingress Controller
`createCommonResources` are shared prerequisites for all Ingress Controllers and they are much broader. I mentioned here https://github.com/nginxinc/nginx-ingress-operator/pull/109#discussion_r647678322 that it changes the logic and it is  not related to the SDK upgrade.

> Calling it from SetupWithManager doesn't work because the resources are not ready yet.

what error do you get?

Reply:
`"error getting ClusterRole: the cache is not started, can not read objects"`

Reply:
@lucacome To read the objects from Kubernetes API, the existing code uses `mgr.GetAPIReader()`. The new code uses `mgr.GetClient()`, that's why you see the error. Note that `mgr.GetAPIReader()` doesn't require caches to start.

The existing code:
https://github.com/nginxinc/nginx-ingress-operator/blob/2c447946710e952d545c7931fac2d13ab9563c59/pkg/controller/nginxingresscontroller/nginxingresscontroller_controller.go#L122-L123